ouroboros-consensus/src/Data/Map/Diff/Strict.hs

```diff
@@ -0,0 +1,137 @@
+{-# LANGUAGE DataKinds                  #-}
+{-# LANGUAGE DeriveAnyClass             #-}
+{-# LANGUAGE DeriveFunctor              #-}
+{-# LANGUAGE DeriveGeneric              #-}
+{-# LANGUAGE DerivingStrategies         #-}
+{-# LANGUAGE GADTs                      #-}
+{-# LANGUAGE GeneralizedNewtypeDeriving #-}
+{-# LANGUAGE KindSignatures             #-}
+{-# LANGUAGE LambdaCase                 #-}
+{-# LANGUAGE MultiParamTypeClasses      #-}
+{-# LANGUAGE StandaloneDeriving         #-}
+
+module Data.Map.Diff.Strict (
+    Diff (..)
+  , DiffEntry (..)
+  , DiffHistory (..)
+  , singletonDelete
+  , singletonInsert
+  ) where
+
+import           Data.Group
+import qualified Data.Map.Merge.Strict as Merge
+import           Data.Map.Strict (Map)
+import           Data.Sequence (Seq (..))
+import qualified Data.Sequence as Seq
+import           GHC.Generics (Generic)
+import           NoThunks.Class (NoThunks)
+
+{------------------------------------------------------------------------------
+  General-purposes diffs for key-value stores
+------------------------------------------------------------------------------}
+
+-- | A diff for key-value stores.
+newtype Diff k v = Diff (Map k (DiffHistory v))
+  deriving stock (Generic, Show, Eq)
+  deriving anyclass (NoThunks)
+
+-- | A history of changes to a value in a key-value store.
+--
+-- A history has an implicit sense of ordering according to time: from left to
+-- right. This means that the left-most element in the history is the
+-- /earliest/ change, while the right-most element in the history is the
+-- /latest/ change.
+newtype DiffHistory v = DiffHistory (Seq (DiffEntry v))
+  deriving stock (Generic, Show, Eq)
+  deriving anyclass (NoThunks)
+
+singleton :: DiffEntry v -> DiffHistory v
+singleton = DiffHistory . Seq.singleton
+
+singletonInsert :: v -> DiffHistory v
+singletonInsert = singleton . Insert
+
+singletonDelete :: v -> DiffHistory v
+singletonDelete = singleton . Delete
+
+-- | A change to a value in a key-value store.
+--
+-- Note: updates are equivalent to inserts, since we consider them to
+-- overwrite previous values.
+data DiffEntry v = Insert !v | Delete !v
+  deriving stock (Generic, Show, Eq, Functor)
+  deriving anyclass (NoThunks)
+
+{------------------------------------------------------------------------------
+  Class instances for @'Diff'@
+------------------------------------------------------------------------------}
+
+instance (Ord k, Eq v) => Semigroup (Diff k v) where
+  Diff m1 <> Diff m2 = Diff $
+    Merge.merge
+      Merge.preserveMissing
+      Merge.preserveMissing
+      (Merge.zipWithMaybeMatched
+        (\_k h1 h2 -> isNonEmptyHistory $ h1 <> h2)
+      )
+      m1
+      m2
```

Comment:
Document why this is how we merge diffs. For example, why do we use `isNonEmptyHistory`?

Reply:
Elaborated on the use of `isNonEmptyHistory` in a comment.